hail/src/main/resources/include/hail/Decoder.h

```diff
@@ -209,6 +211,50 @@ class LEB128InputBuffer {
     bool read_boolean() { return buf_.read_boolean(); }
 };
 
+template<typename Decoder>
+class ReadIterator {
+private:
+  std::shared_ptr<Decoder> dec_;
+  Region * region_;
+  NativeStatus * st_;
+  char * value_;
+
+public:
+  ReadIterator() :
+  dec_(nullptr),
+  region_(nullptr),
+  st_(nullptr),
+  value_(nullptr) { }
+
+  ReadIterator(std::shared_ptr<Decoder> dec, Region * region, NativeStatus* st) :
+  dec_(dec),
+  region_(region),
+  st_(st),
+  value_(nullptr) { }
+
+  ReadIterator<Decoder>& operator++() {
+    if (dec_ != nullptr) {
+      if (dec_->decode_byte(st_)) {
+        value_ = dec_->decode_row(st_, region_);
+      } else {
+        value_ = nullptr;
+        dec_ = nullptr;
+      }
+    }
+    return *this;
+  }
+
+  bool operator==(ReadIterator other) const {
```

Comment:
Binary operations like comparison should usually be defined as free functions, friends of the class if needed. In this case I would write
```c++
template<typename Decoder>
friend bool operator==(ReadIterator<Decoder> const& lhs, ReadIterator<Decoder> const& rhs) {
  return (lhs._dec == rhs._dec) && (lhs.value_ == rhs.value_)
}
```
This ensures equality is symmetric.